#include "mainwindow.h"
#include "./ui_mainwindow.h"
#include "tablemanager.h"
#include "filemanager.h"
#include "chartmanager.h"
#include "ChartDialog.h"
#include "searchmanager.h"

#include <QFileDialog>
#include <QMessageBox>
#include <QDialog>
#include <QFormLayout>
#include <QLineEdit>
#include <QDateEdit>
#include <QComboBox>
#include <QDialogButtonBox>
#include <QDoubleValidator>
#include <QIntValidator>
#include <QRegularExpressionValidator>
#include <QRegularExpression>
#include <QMap>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow)

{
    ui->setupUi(this);

    m_tableManager = new TableManager(ui->insuranceTable, this, this);
    m_fileManager = new FileManager();
    m_chartManager = new ChartManager(ui->chartWidget_, this);
    searchManager = new SearchManager(this);

    m_tableManager->initializeTable();

    columnMap.clear();
    QTableWidget *currentTable = m_tableManager->getTableWidget();
    if (currentTable)
    {
        for (int col = 0; col < currentTable->columnCount(); ++col)
        {
            QTableWidgetItem *headerItem = currentTable->horizontalHeaderItem(col);
            if (headerItem)
            {
                QString name = headerItem->text().trimmed().toLower();
                columnMap[name] = col;
            }
        }

        QMap<QString, QString> aliases;
        aliases["–∑–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω—ã–π"] = "—Ñ–∏–æ –∑–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω–æ–≥–æ";
        aliases["–∞–≥–µ–Ω—Ç"] = "—Ñ–∏–æ –∞–≥–µ–Ω—Ç–∞";
        aliases["—Å–∫–∏–¥–∫–∞"] = "—Å–∫–∏–¥–∫–∞";
        aliases["—Ç–µ–ª–µ—Ñ–æ–Ω"] = "—Ç–µ–ª–µ—Ñ–æ–Ω";
        aliases["–¥–∞—Ç–∞"] = "–¥–∞—Ç–∞ —Å—Ç—Ä–∞—Ö–æ–≤–∫–∏";
        aliases["–¥–∞—Ç–∞ –æ–∫–æ–Ω—á–∞–Ω–∏—è"] = "–¥–∞—Ç–∞ –æ–∫–æ–Ω—á–∞–Ω–∏—è";

        for (auto it = aliases.constBegin(); it != aliases.constEnd(); ++it)
        {
            const QString &alias = it.key();
            const QString &actualHeaderName = it.value();
            if (columnMap.contains(actualHeaderName))
            {
                columnMap[alias] = columnMap[actualHeaderName];
            }
        }
    }

    ui->searchLineEdit->setPlaceholderText("üîç –ø–∞—Ä–∞–º–µ—Ç—Ä: —Ç–µ–∫—Å—Ç");

    setupConnections();

    m_chartManager->updatePieChart(m_tableManager->getTableWidget());
}

MainWindow::~MainWindow()
{
    delete ui;
    delete m_fileManager;
}

void MainWindow::setupConnections()
{
    connect(ui->openChartButton, &QPushButton::clicked, this, &MainWindow::on_openChartsButton_clicked);
    connect(ui->searchLineEdit, &QLineEdit::textChanged, this, [=](const QString &text)
            { searchManager->search(m_tableManager->getTableWidget(), text, columnMap); });
}

void MainWindow::on_loadButton_clicked()
{
    QString fileName = QFileDialog::getOpenFileName(this, "–í—ã–±—Ä–∞—Ç—å —Ñ–∞–π–ª –¥–ª—è –∑–∞–≥—Ä—É–∑–∫–∏",
                                                    "", "Text Files (*.txt);;All Files (*)");
    if (!fileName.isEmpty())
    {
        bool success = m_fileManager->loadDataFromFile(fileName, m_tableManager, this);
        if (success)
        {
            m_chartManager->updatePieChart(m_tableManager->getTableWidget());
        }
    }
}

void MainWindow::handleCellChanged(int row, int column)
{

    bool needsChartUpdate = false;

    if (row >= 0 && column == -1)
    {

        if (validateRowData(row))
        {
            needsChartUpdate = true;
        }
        else
        {
        }
    }

    else if (column == Column::Type || column == Column::StartDate || column == Column::EndDate || (row == -1 && column == -1))
    {
        needsChartUpdate = true;
    }

    if (needsChartUpdate)
    {
        m_chartManager->updatePieChart(m_tableManager->getTableWidget());
    }
}

void MainWindow::on_addButton_clicked()
{

    QDialog dialog(this);
    dialog.setWindowTitle("–î–æ–±–∞–≤–ª–µ–Ω–∏–µ –Ω–æ–≤–æ–π –∑–∞–ø–∏—Å–∏");

    QFormLayout form(&dialog);

    QLineEdit *insuredFioEdit = new QLineEdit(&dialog);
    insuredFioEdit->setPlaceholderText("–§–∞–º–∏–ª–∏—è –ò–º—è –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –§–∞–º–∏–ª–∏—è –ò–º—è -)");

    QLineEdit *agentFioEdit = new QLineEdit(&dialog);
    agentFioEdit->setPlaceholderText("–§–∞–º–∏–ª–∏—è –ò–º—è –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –§–∞–º–∏–ª–∏—è –ò–º—è -)");

    QLineEdit *amountEdit = new QLineEdit(&dialog);
    amountEdit->setValidator(new QDoubleValidator(0, 1e9, 2, &dialog));
    amountEdit->setPlaceholderText("–ù–∞–ø—Ä–∏–º–µ—Ä, 15000.50");

    QDateEdit *startDateEdit = new QDateEdit(QDate::currentDate(), &dialog);
    startDateEdit->setCalendarPopup(true);
    startDateEdit->setDisplayFormat("yyyy-MM-dd");

    QDateEdit *endDateEdit = new QDateEdit(QDate::currentDate().addYears(1), &dialog);
    endDateEdit->setCalendarPopup(true);
    endDateEdit->setDisplayFormat("yyyy-MM-dd");

    QComboBox *typeCombo = new QComboBox(&dialog);
    typeCombo->addItems({"–ê–≤—Ç–æ", "–ó–¥–æ—Ä–æ–≤—å–µ", "–ñ–∏–∑–Ω—å", "–ò–Ω–æ–µ"});
    QLineEdit *phoneEdit = new QLineEdit(&dialog);
    phoneEdit->setPlaceholderText("+7(XXX)XXX-XX-XX –∏–ª–∏ 7XXXXXXXXXX");

    QLineEdit *discountEdit = new QLineEdit("0", &dialog);

    discountEdit->setValidator(new QIntValidator(0, 100, &dialog));
    discountEdit->setPlaceholderText("0-100");

    form.addRow("–§–ò–û –ó–∞—Å—Ç—Ä–∞—Ö.*:", insuredFioEdit);
    form.addRow("–§–ò–û –ê–≥–µ–Ω—Ç–∞*:", agentFioEdit);
    form.addRow("–°—É–º–º–∞*:", amountEdit);
    form.addRow("–î–∞—Ç–∞ –°—Ç—Ä–∞—Ö–æ–≤–∫–∏:", startDateEdit);
    form.addRow("–î–∞—Ç–∞ –û–∫–æ–Ω—á–∞–Ω–∏—è:", endDateEdit);
    form.addRow("–í–∏–¥ –°—Ç—Ä–∞—Ö–æ–≤–∫–∏:", typeCombo);
    form.addRow("–¢–µ–ª–µ—Ñ–æ–Ω*:", phoneEdit);
    form.addRow("–°–∫–∏–¥–∫–∞ (%):", discountEdit);

    QDialogButtonBox buttonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel,
                               Qt::Horizontal, &dialog);
    form.addRow(&buttonBox);

    connect(&buttonBox, &QDialogButtonBox::accepted, &dialog, &QDialog::accept);
    connect(&buttonBox, &QDialogButtonBox::rejected, &dialog, &QDialog::reject);

    if (dialog.exec() == QDialog::Accepted)
    {
        QString errorMsg;

        QString insuredFioRaw = insuredFioEdit->text().trimmed();
        QStringList insuredParts = insuredFioRaw.split(' ', Qt::SkipEmptyParts);
        if (insuredFioRaw.isEmpty())
        {
            errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω–æ –§–ò–û –ó–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω–æ–≥–æ.\n";
        }
        else if (insuredParts.count() != 3)
        {
            errorMsg += "–§–ò–û –ó–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω–æ–≥–æ: –£–∫–∞–∂–∏—Ç–µ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –ø—Ä–æ—á–µ—Ä–∫ '-' –≤–º–µ—Å—Ç–æ –æ—Ç—á–µ—Å—Ç–≤–∞).\n";
        }

        QString agentFioRaw = agentFioEdit->text().trimmed();
        QStringList agentParts = agentFioRaw.split(' ', Qt::SkipEmptyParts);
        if (agentFioRaw.isEmpty())
        {
            errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω–æ –§–ò–û –ê–≥–µ–Ω—Ç–∞.\n";
        }
        else if (agentParts.count() != 3)
        {
            errorMsg += "–§–ò–û –ê–≥–µ–Ω—Ç–∞: –£–∫–∞–∂–∏—Ç–µ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –ø—Ä–æ—á–µ—Ä–∫ '-' –≤–º–µ—Å—Ç–æ –æ—Ç—á–µ—Å—Ç–≤–∞).\n";
        }

        if (amountEdit->text().trimmed().isEmpty())
        {
            errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω–∞ –°—É–º–º–∞.\n";
        }
        else
        {
            bool amountOk;
            amountEdit->text().toDouble(&amountOk);
            if (!amountOk)
            {
                errorMsg += "–°—É–º–º–∞ —É–∫–∞–∑–∞–Ω–∞ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ (–æ–∂–∏–¥–∞–µ—Ç—Å—è —á–∏—Å–ª–æ).\n";
            }
        }

        bool discountOk;
        int discountVal = discountEdit->text().toInt(&discountOk);
        if (!discountOk || discountVal < 0 || discountVal > 100)
        {

            errorMsg += "–°–∫–∏–¥–∫–∞ —É–∫–∞–∑–∞–Ω–∞ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ (–æ–∂–∏–¥–∞–µ—Ç—Å—è —Ü–µ–ª–æ–µ —á–∏—Å–ª–æ –æ—Ç 0 –¥–æ 100).\n";
        }

        if (startDateEdit->date() > endDateEdit->date())
        {
            errorMsg += "–î–∞—Ç–∞ –æ–∫–æ–Ω—á–∞–Ω–∏—è –Ω–µ –º–æ–∂–µ—Ç –±—ã—Ç—å —Ä–∞–Ω—å—à–µ –¥–∞—Ç—ã –Ω–∞—á–∞–ª–∞ —Å—Ç—Ä–∞—Ö–æ–≤–∫–∏.\n";
        }

        QString phoneRaw = phoneEdit->text().trimmed();
        if (phoneRaw.isEmpty())
        {
            errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω –¢–µ–ª–µ—Ñ–æ–Ω.\n";
        }
        else
        {
            QString digitsOnly = phoneRaw;
            digitsOnly.remove(QRegularExpression("\\D"));
            if (digitsOnly.length() != 11)
            {
                errorMsg += "–¢–µ–ª–µ—Ñ–æ–Ω: –î–æ–ª–∂–Ω–æ –±—ã—Ç—å 11 —Ü–∏—Ñ—Ä (–Ω–∞–ø—Ä–∏–º–µ—Ä, 79991234567).\n";
            }
            else if (digitsOnly.at(0) != QChar('7'))
            {
                errorMsg += "–¢–µ–ª–µ—Ñ–æ–Ω: –î–æ–ª–∂–µ–Ω –Ω–∞—á–∏–Ω–∞—Ç—å—Å—è —Å 7 (–∏–ª–∏ +7).\n";
            }
        }

        if (!errorMsg.isEmpty())
        {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞ –≤–≤–æ–¥–∞", errorMsg);
            return;
        }

        QStringList fields;
        fields << insuredParts;
        fields << agentParts;
        fields << amountEdit->text().trimmed();
        fields << startDateEdit->date().toString("yyyy-MM-dd");
        fields << endDateEdit->date().toString("yyyy-MM-dd");
        fields << typeCombo->currentText();
        fields << phoneRaw;
        fields << discountEdit->text().trimmed() + "%";

        int newRow = ui->insuranceTable->rowCount();
        m_tableManager->addTableRow(newRow, fields, true);

        m_chartManager->updatePieChart(m_tableManager->getTableWidget());
    }
}

void MainWindow::on_deleteButton_clicked()
{
    m_tableManager->deleteSelectedRow();
    m_chartManager->updatePieChart(m_tableManager->getTableWidget());
}

void MainWindow::on_exportButton_clicked()
{
    QString fileName = QFileDialog::getSaveFileName(this, "–≠–∫—Å–ø–æ—Ä—Ç –¥–∞–Ω–Ω—ã—Ö",
                                                    "", "–¢–µ–∫—Å—Ç–æ–≤—ã–µ —Ñ–∞–π–ª—ã (*.txt);;–í—Å–µ —Ñ–∞–π–ª—ã (*)");
    if (!fileName.isEmpty())
    {
        m_fileManager->exportDataToFile(fileName, m_tableManager, this);
    }
}

void MainWindow::on_openChartsButton_clicked()
{
    QMap<QString, double> agentSums = m_tableManager->getAgentSums();
    QMap<QString, int> monthCounts = m_tableManager->getMonthCounts();

    ChartDialog *dialog = new ChartDialog(this);
    dialog->setData(agentSums, monthCounts);
    dialog->setWindowTitle("–ê–Ω–∞–ª–∏—Ç–∏–∫–∞ —Å—Ç—Ä–∞—Ö–æ–≤—ã—Ö —Å–¥–µ–ª–æ–∫");
    dialog->exec();
    delete dialog;
}

bool MainWindow::validateRowData(int row)
{
    if (!m_tableManager || !m_tableManager->getTableWidget())
        return false;
    QTableWidget *table = m_tableManager->getTableWidget();
    if (row < 0 || row >= table->rowCount())
        return false;

    QString errorMsg;
    bool isValid = true;

    QString insuredFioStr = table->item(row, Column::InsuredFio) ? table->item(row, Column::InsuredFio)->text().trimmed() : "";
    QString agentFioStr = table->item(row, Column::AgentFio) ? table->item(row, Column::AgentFio)->text().trimmed() : "";
    QString amountStr = table->item(row, Column::Amount) ? table->item(row, Column::Amount)->text().trimmed() : "";
    QString phoneStr = table->item(row, Column::Phone) ? table->item(row, Column::Phone)->text().trimmed() : "";
    QString discountStr = table->item(row, Column::Discount) ? table->item(row, Column::Discount)->text().trimmed() : "0%";
    QDate startDate = QDate();
    if (auto *de = qobject_cast<QDateEdit *>(table->cellWidget(row, Column::StartDate)))
        startDate = de->date();
    QDate endDate = QDate();
    if (auto *de = qobject_cast<QDateEdit *>(table->cellWidget(row, Column::EndDate)))
        endDate = de->date();

    QStringList insuredParts = insuredFioStr.split(' ', Qt::SkipEmptyParts);
    if (insuredFioStr.isEmpty())
    {
        errorMsg += "–§–ò–û –ó–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω–æ–≥–æ: –Ω–µ –∑–∞–ø–æ–ª–Ω–µ–Ω–æ.\n";
        isValid = false;
    }
    else if (insuredParts.count() != 3)
    {
        errorMsg += "–§–ò–û –ó–∞—Å—Ç—Ä–∞—Ö–æ–≤–∞–Ω–Ω–æ–≥–æ: –£–∫–∞–∂–∏—Ç–µ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –ø—Ä–æ—á–µ—Ä–∫ '-' –≤–º–µ—Å—Ç–æ –æ—Ç—á–µ—Å—Ç–≤–∞).\n";
        isValid = false;
    }

    QStringList agentParts = agentFioStr.split(' ', Qt::SkipEmptyParts);
    if (agentFioStr.isEmpty())
    {
        errorMsg += "–§–ò–û –ê–≥–µ–Ω—Ç–∞: –Ω–µ –∑–∞–ø–æ–ª–Ω–µ–Ω–æ.\n";
        isValid = false;
    }
    else if (agentParts.count() != 3)
    {
        errorMsg += "–§–ò–û –ê–≥–µ–Ω—Ç–∞: –£–∫–∞–∂–∏—Ç–µ –§–∞–º–∏–ª–∏—é, –ò–º—è –∏ –û—Ç—á–µ—Å—Ç–≤–æ (–∏–ª–∏ –ø—Ä–æ—á–µ—Ä–∫ '-' –≤–º–µ—Å—Ç–æ –æ—Ç—á–µ—Å—Ç–≤–∞).\n";
        isValid = false;
    }

    if (amountStr.isEmpty())
    {
        errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω–∞ –°—É–º–º–∞.\n";
        isValid = false;
    }
    else
    {
        bool amountOk;
        double amount = amountStr.toDouble(&amountOk);
        if (!amountOk || amount < 0)
        {
            errorMsg += "–°—É–º–º–∞ –¥–æ–ª–∂–Ω–∞ –±—ã—Ç—å –ø–æ–ª–æ–∂–∏—Ç–µ–ª—å–Ω—ã–º —á–∏—Å–ª–æ–º.\n";
            isValid = false;
        }
    }

    if (phoneStr.isEmpty())
    {
        errorMsg += "–ù–µ —É–∫–∞–∑–∞–Ω –¢–µ–ª–µ—Ñ–æ–Ω.\n";
        isValid = false;
    }
    else
    {
        QString digitsOnly = phoneStr;
        digitsOnly.remove(QRegularExpression("\\D"));
        if (digitsOnly.length() != 11)
        {
            errorMsg += "–¢–µ–ª–µ—Ñ–æ–Ω: –û–∂–∏–¥–∞–µ—Ç—Å—è 11 —Ü–∏—Ñ—Ä (–Ω–∞–ø—Ä–∏–º–µ—Ä, +7(999)123-45-67).\n";
            isValid = false;
        }
        else if (!digitsOnly.startsWith('7'))
        {
            errorMsg += "–¢–µ–ª–µ—Ñ–æ–Ω: –î–æ–ª–∂–µ–Ω –Ω–∞—á–∏–Ω–∞—Ç—å—Å—è —Å 7 (–∏–ª–∏ +7).\n";
            isValid = false;
        }
    }

    if (!isValid)
    {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞ –¥–∞–Ω–Ω—ã—Ö –≤ —Å—Ç—Ä–æ–∫–µ " + QString::number(row + 1), errorMsg);
    }

    return isValid;
}
